File: 5.双指针法/1.移除元素/283.移动零.cpp
/*
 * @lc app=leetcode.cn id=283 lang=cpp
 *
 * [283] 移动零
 */
#include <bits/stdc++.h>
using namespace std;
// @lc code=start
class Solution {
public:
    void moveZeroes(vector<int>& nums) {
        int i = 0, j = 1;
        while (j < nums.size())
        {
            if (nums[i] == 0 && nums[j] != 0)
            {
                swap(nums[i ++], nums[j ++]);
            }
            else if (nums[i] == 0 && nums[j] == 0)
            {
                j ++;
            }
            else if (nums[i] != 0 && nums[j] == 0)
            {
                i ++, j ++;
            }
            else
            {
                i ++, j ++;
            }
        }
        return;
    }
};
// @lc code=end

File: HOT100/2.2/283.移动零.py
#
# @lc app=leetcode.cn id=283 lang=python3
#
# [283] 移动零
#
from mytools import *
# @lc code=start
class Solution:
    def moveZeroes(self, nums: List[int]) -> None:
        """
        Do not return anything, modify nums in-place instead.
        """
        n = len(nums)
        i = 0
        for j in range(n):
            if nums[j] != 0:
                nums[i] = nums[j]
                i += 1
        for k in range(i, n): nums[k] = 0
# @lc code=end

class Solution:
    def moveZeroes(self, nums: List[int]) -> None:
        """
        Do not return anything, modify nums in-place instead.
        """
        n = len(nums)
        for i in range(n):
            if nums[i] == 0: break
        sta = i
        for j in range(sta + 1, n):
            if nums[j] == 0: continue
            while i < j and nums[i] != 0: i += 1
            nums[i], nums[j] = nums[j], nums[i]